### PR TITLE
handle diags that happen during repl code completion

### PR DESCRIPTION
**Pull request by marcrasi**
If a REPL completion request encounters an error during typechecking and produces an AST with an error type, then the ASTVerifier must not run, because the ASTVerifier is only supposed to run on ASTs without errors.

REPLCodeCompletion.cpp's previous strategy for suppressing diagnostics stopped the diagnostics from being added to the ASTContext, so ASTVerifier did not know that there were errors, so it tried to run.

This PR makes it use DiagnosticSuppression, which suppresses the diagnostic consumer that prints diagnostics, but which does not stop the diagnostics from being added to the ASTContext.

Related discussions in: #22580 and #22579

### CODE REVIEW

**Comment by jrose-apple** [comment]
This makes more sense, although it's still pretty brute-force! Thanks, @marcrasi.

**Review by xedin — APPROVED** [review]
LGTM! Thank you!

**Comment by xedin** [comment]
@swift-ci please smoke test

**Comment by marcrasi** [comment]
While playing with this a bit, I realized that I now need to reset the error state.

A concrete problem from not resetting it: in the integrated repl, ask for a completion and then hit enter to execute the line. It'll silently not execute the line because it thinks that there is an error.

I uploaded a commit that resets the error state.

**Comment by marcrasi** [comment]
@swift-ci please smoke test

**Comment by marcrasi** [comment]
@swift-ci please smoke test

File: lib/IDE/REPLCodeCompletion.cpp
Comment (jrose-apple) [inline lib/IDE/REPLCodeCompletion.cpp]: Ah, this isn't correct. There might have been errors *before* we started, right?

Comment (marcrasi) [reply lib/IDE/REPLCodeCompletion.cpp]: My guess is that there is never any intentional error state that must be preserved across completion calculations. Of course, there could be some accidental error state from other operations that forgot to reset the error state.

The reason I think this is that I wouldn't expect there to be any "ongoing compilation" happening when this function executes. The only compilations I know about are:
* When you execute a line in the REPL, it compiles that line. Errors from these executions become irrelevant once the execution returns.
* When you ask for a completion, it compiles your code. Errors from these compilations become irrelevant once the completion request returns.

I don't have much experience with any of this code, so I could be totally wrong.

Comment (jrose-apple) [reply lib/IDE/REPLCodeCompletion.cpp]: I…guess that's probably the case. @dcci, @adrian-prantl, @jimingham, does that sound right for LLDB's REPL? (The integrated REPL is not important.)

Comment (marcrasi) [reply lib/IDE/REPLCodeCompletion.cpp]: Ping! Should I wait for an answer to this before merging?

Comment (adrian-prantl) [reply lib/IDE/REPLCodeCompletion.cpp]: I don't know the answer to this. If no one else objects let's give this a try.

Comment (marcrasi) [reply lib/IDE/REPLCodeCompletion.cpp]: Seems like there have been no objections yet. I'm going to give it a try!

**Comment by marcrasi** [comment]
@swift-ci please smoke test

**Comment by marcrasi** [comment]
@swift-ci Please smoke test OS X platform

**Pull request merged**
